kv/dist_sender.go

```diff
@@ -406,90 +379,141 @@ func (ds *DistSender) sendRPC(trace *tracer.Trace, rangeID proto.RangeID, replic
 	return replies[0].(proto.Response), nil
 }
 
-// getDescriptors takes a call and looks up the corresponding range
-// descriptors associated with it. First, the range descriptor for
-// call.Args.Key is looked up. If call.Args.EndKey exceeds that of the
-// returned descriptor, the next descriptor is obtained as well.
-func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *proto.RangeDescriptor, error) {
-	// If this is a PushTxn, set ignoreIntents option as
-	// necessary. This prevents a potential infinite loop; see the
-	// comments in proto.RangeLookupRequest.
-	options := lookupOptions{}
-	if pushArgs, ok := call.Args.(*proto.PushTxnRequest); ok {
-		options.ignoreIntents = pushArgs.RangeLookup
-	}
-
+// getDescriptors looks up the range descriptor to use for a query over the
+// key range [from,to), with the given lookupOptions. The range descriptor
+// which contains the range in which the request should start its query is
+// returned first; a second one is optionally returned in case the given
+// range reaches outside the first descriptor.
+// In case either of the descriptors is discovered stale, the returned closure
+// should be called; it evicts the cache appropriately.
+// Note that `from` and `to` are not necessarily Key and EndKey from a
+// RequestHeader; it's assumed that they've been translated to key addresses
+// already (via KeyAddress).
+func (ds *DistSender) getDescriptors(from, to proto.Key, options lookupOptions) (*proto.RangeDescriptor, *proto.RangeDescriptor, func(), error) {
 	var desc *proto.RangeDescriptor
 	var err error
-	_, isReverseScan := call.Args.(*proto.ReverseScanRequest)
-	if !isReverseScan {
-		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().Key, options)
+	var descKey proto.Key
+	if !options.useReverseScan {
+		descKey = from
 	} else {
-		options.useReverseScan = true
-		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().EndKey, options)
+		descKey = to
 	}
+	desc, err = ds.rangeCache.LookupRangeDescriptor(descKey, options)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// Checks whether need to get next range descriptor. If so, returns true
 	// and the key to look up, depending on whether we're in reverse mode.
-	needAnother := func(desc *proto.RangeDescriptor, isReverseScan bool) (proto.Key, bool) {
-		if isReverseScan {
-			return desc.StartKey, call.Args.Header().Key.Less(desc.StartKey)
+	needAnother := func(desc *proto.RangeDescriptor, isReverse bool) (proto.Key, bool) {
+		if isReverse {
+			return desc.StartKey, from.Less(desc.StartKey)
```

Comment:
no, the end key is exclusive, so if you're reverse scanning `[a,c)` (which starts at `c`) and your desc is `[b,c)`, you want your next iteration to look up `b` (in reverse mode, so that you'll go for the range with `EndKey=b`: `[a,b)`.)